src/storage-types/src/sources/load_generator.rs

```diff
@@ -174,6 +174,19 @@ pub fn external_reference_to_output(
     }
 }
 
+/// Returns the view of a load-generator source given an external reference
+/// from a `CREATE TABLE .. FROM SOURCE` statement, which is generated during
+/// purification such that the external reference is (LoadGenerator::schema_name(), view_name)
+pub fn table_reference_to_output(external_reference: &UnresolvedItemName) -> LoadGeneratorOutput {
+    let inner = &external_reference.0;
```

Comment:
What happens if the user specifies three components in this reference where the first one is a "database"? Every PR makes me increasingly nervous that we will mishandle them somewhere along the line. We need to enforce that when a user types an external reference we convert it to its canonical representation "at the edges", i.e during purification and as early as possible in there.

 From that point onwards we ought to have a different type, call it `ExternalReference` that is *guaranteed* to have a `.schema()` and a `.name()` method that return the correct data. Sources should advertise lists of available `ExternalReferences` and comparisons should almost never attempt to look into individual components. The only thing that needs to do that is the initial resolution during purification.

Reply:
> What happens if the user specifies three components in this reference where the first one is a "database"?

That happens during purification -- this `table_reference_to_output ` method is operating on the purified / normalized 'external reference' normalized during purification. It's separate from the `subsource_reference_to_output` method above because there is no need to store a fake database name on the purified statement and since we're implementing a completely new statement I thought I might as well clean that up such that the fake db name can be removed once we get rid of subsources.

> We need to enforce that when a user types an external reference we convert it to its canonical representation "at the edges", i.e during purification and as early as possible in there. 

Yes this is already the case for all sources except load generators at the moment. I was planning to fix load-gens in a future PR by storing the `LoadGeneratorOutput` directly on the `details` protobuf during purification but to do so would need to write a migration for existing subsources and I didn't want to clutter this PR. Doing this would allow us to remove these two methods here since we would never need to look at external-references again after purification, which I agree is a magnet for a potential future bug. I'll push a commit to this PR that does this refactor and migration which means we won't need this method at all.

> From that point onwards we ought to have a different type, call it ExternalReference that is guaranteed to have a .schema() and a .name() method that return the correct data. Sources should advertise lists of available ExternalReferences and comparisons should almost never attempt to look into individual components. The only thing that needs to do that is the initial resolution during purification.

We don't need to add that machinery -- we already do something similar through `SourceExportStatementDetails` -- I just need to convert load-gen sources to use it. MySQL and Postgres don't look at their 'external reference' after purification at all -- the upstream info is instead encoded into the `details` protobuf in the source-specific form:
https://github.com/MaterializeInc/materialize/blob/e194a6f3fdd59c8efd5bd7521da7fba9cb79ac3d/src/storage-types/src/sources.proto#L82-L91

Reply:
Okay I've just pushed a new commit that converts load generators to use the same semantics as MySQL and Postgres sources -- external references are resolved only during purification and used to populate the `details` option on each statement, which is then used in planning and other downstream dependencies. The external reference `UnresolvedItemName` is not needed at all after purification:
https://github.com/MaterializeInc/materialize/pull/28954/commits/725dbb384677ad938d2f4c245dbbbbdbb5e8b937

Reply:
Also to see where this is going -- this draft PR https://github.com/MaterializeInc/materialize/pull/29268 refactors things such that 'external references' aren't provided to the storage layer at all -- they are unnecessary after purification happens